README.md
# rbcf
R Bindings for htslib/bcf+vcf

[![Build Status](https://travis-ci.org/lindenb/rbcf.svg)](https://travis-ci.org/lindenb/rbcf)

![Last commit](https://img.shields.io/github/last-commit/lindenb/rbcf.png)

## INSTALL

Requirements: R, git, curl-dev, 

```
git clone "https://github.com/lindenb/rbcf"

cd rbcf

make
```

## Author

Pierre Lindenbaum PhD. Institut du Thorax. Nantes, France.
@yokofakun

## Examples
### Show Htslib and Rbcf versions

**Code**:

```
# load the library
library(rbcf)
#print the version of the associated htslib 
paste("HTSLIB:",htslib.version())
#print the version of rbcf
paste("RBCF:",rcbf.version())

```

**Output**:

```
[1] "HTSLIB: 1.10.2"
[1] "RBCF: 0.0-1"
```


### Open and close a VCF file

**Code**:

```
# load rbcf
library(rbcf)
# we don't need the index for this file
fp <- bcf.open("./data/rotavirus_rf.01.vcf",FALSE)
# error (exit 0 for tests)
if(is.null(fp)) quit(save="no",status=0,runLast=FALSE)
# dispose the vcf reader
bcf.close(fp)
print("Done.")

```

**Output**:

```
[1] TRUE
[1] "Done."
```


### Print the INFOs in the VCF header

**Code**:

```
# load rbcf
library(rbcf)
# we don't need the index for this file
fp <- bcf.open("./data/rotavirus_rf.01.vcf",FALSE)
# error on opening (exit 0 for tests)
if(is.null(fp)) quit(save="no",status=0,runLast=FALSE)
# print INFO
bcf.infos(fp)
# dispose the vcf reader
bcf.close(fp)
# print the table
```

**Output**:

```
         ID Number    Type
INDEL INDEL      0    Flag
IDV     IDV      1 Integer
IMF     IMF      1   Float
DP       DP      1 Integer
VDB     VDB      1   Float
RPB     RPB      1   Float
MQB     MQB      1   Float
BQB     BQB      1   Float
MQSB   MQSB      1   Float
SGB     SGB      1   Float
MQ0F   MQ0F      1   Float
ICB     ICB      1   Float
HOB     HOB      1   Float
AC       AC      A Integer
AN       AN      1 Integer
DP4     DP4      4 Integer
MQ       MQ      1 Integer
                                                                                         Descri(...)
INDEL                                                      "Indicates that the variant is an IN(...)
IDV                                                    "Maximum number of reads supporting an i(...)
IMF                                                  "Maximum fraction of reads supporting an i(...)
DP                                                                                  "Raw read d(...)
VDB   "Variant Distance Bias for filtering splice-site artefacts in RNA-seq data (bigger is bet(...)
RPB                                   "Mann-Whitney U test of Read Position Bias (bigger is bet(...)
MQB                                 "Mann-Whitney U test of Mapping Quality Bias (bigger is bet(...)
BQB                                    "Mann-Whitney U test of Base Quality Bias (bigger is bet(...)
MQSB                      "Mann-Whitney U test of Mapping Quality vs Strand Bias (bigger is bet(...)
SGB                                                                      "Segregation based met(...)
MQ0F                                                     "Fraction of MQ0 reads (smaller is bet(...)
ICB                                        "Inbreeding Coefficient Binomial test (bigger is bet(...)
HOB                                          "Bias in the number of HOMs number (smaller is bet(...)
AC                      "Allele count in genotypes for each ALT allele, in the same order as li(...)
AN                                                     "Total number of alleles in called genot(...)
DP4            "Number of high-quality ref-forward , ref-reverse, alt-forward and alt-reverse b(...)
MQ                                                                         "Average mapping qua(...)
[1] TRUE
```


### Print the FORMATs in the VCF header

**Code**:

```
# load rbcf
library(rbcf)
# we don't need the index for this file
fp <- bcf.open("./data/rotavirus_rf.01.vcf",FALSE)
# error on opening (exit 0 for tests)
if(is.null(fp)) quit(save="no",status=0,runLast=FALSE)
# print FORMAT
bcf.formats(fp)
# dispose the vcf reader
bcf.close(fp)
```

**Output**:

```
   ID Number    Type                                 Description
PL PL      G Integer "List of Phred-scaled genotype likelihoods"
GT GT      1  String                                  "Genotype"
[1] TRUE
```


### Print the FILTERs in the VCF header

**Code**:

```
# load rbcf
library(rbcf)
# we don't need the index for this file
fp <- bcf.open("./data/gnomad.exomes.r2.0.1.sites.bcf",FALSE)
# error on opening (exit 0 for tests)
if(is.null(fp)) quit(save="no",status=0,runLast=FALSE)
# print FILTERs
bcf.filters(fp)
# dispose the vcf reader
bcf.close(fp)

```

**Output**:

```
                             ID
PASS                       PASS
AC0                         AC0
InbreedingCoeff InbreedingCoeff
LCR                         LCR
RF                           RF
SEGDUP                   SEGDUP
                                                                                               (...)
PASS                                                                                           (...)
AC0             "Allele Count is zero (i.e. no high-confidence genotype (GQ >= 20, DP >= 10, AB(...)
InbreedingCoeff                                                                                (...)
LCR                                                                                        "In (...)
RF                                                  "Failed random forests filters (SNV cutoff (...)
SEGDUP                                                                              "In a segme(...)
[1] TRUE
```


### Print the Samples in the VCF header
 The samples are defined in the '\#CHROM' line of the VCF

**Code**:

```
# load rbcf
library(rbcf)
# we don't need the index for this file
fp <- bcf.open("./data/rotavirus_rf.01.vcf",FALSE)
# error on opening (exit 0 for tests)
if(is.null(fp)) quit(save="no",status=0,runLast=FALSE)
# print the number of samples
paste("Number of samples:",bcf.nsamples(fp))
# get the name for the 1st sample
paste("First sample:",bcf.sample.at(fp,1))
# get the 1-based index for the samples
bcf.sample2index(fp,c("S1","S2","S3","missing"))
# get all the samples
bcf.samples(fp)
# dispose the vcf reader
bcf.close(fp)
```

**Output**:

```
[1] "Number of samples: 5"
[1] "First sample: S1"
     S1      S2      S3 missing 
      1       2       3       0 
[1] "S1" "S2" "S3" "S4" "S5"
[1] TRUE
```


### Print the Dictionary in the VCF header

**Code**:

```
# load rbcf
library(rbcf)
# we don't need the index for this file
fp <- bcf.open("./data/rotavirus_rf.01.vcf",FALSE)
# error on opening (exit 0 for tests)
if(is.null(fp)) quit(save="no",status=0,runLast=FALSE)
# print the dictionary
bcf.dictionary(fp)
# dispose the vcf reader
bcf.close(fp)

```

**Output**:

```
     chrom size
RF01  RF01 3302
RF02  RF02 2687
RF03  RF03 2592
RF04  RF04 2362
RF05  RF05 1579
RF06  RF06 1356
RF07  RF07 1074
RF08  RF08 1059
RF09  RF09 1062
RF10  RF10  751
RF11  RF11  666
[1] TRUE
```


### Print the Indexed Chromosomes

**Code**:

```
# load rbcf
library(rbcf)
# Open the indexed VCF
fp <- bcf.open("./data/rotavirus_rf.02.vcf.gz")
# error on opening (exit 0 for tests)
if(is.null(fp)) quit(save="no",status=0,runLast=FALSE)
# get the indexed contigs
bcf.contigs(fp)
# dispose the vcf reader
bcf.close(fp)

```

**Output**:

```
 [1] "RF01" "RF02" "RF03" "RF04" "RF05" "RF06" "RF07" "RF08" "RF09" "RF10"
[11] "RF11"
[1] TRUE
```


### Scanning the variants

**Code**:

```
# load rbcf
library(rbcf)

# create a function counting variants in a VCF
count.variants<-function(filename) {
	# we don't need the index for this file
	fp <- bcf.open(filename,FALSE)
	# error on opening
	if(is.null(fp)) return(-1)
	# number of variants
	n<-0
	# loop while we can read a variant
	while(!is.null(vc<-bcf.next(fp))) {
		# increment the count
		n<-n+1
	}
	# dispose the vcf reader
	bcf.close(fp)
	# return the number of variant
	n
}

# filenames
vcfs<-c(
	"./data/gnomad.exomes.r2.0.1.sites.bcf",
	"./data/rotavirus_rf.01.vcf",
	"./data/rotavirus_rf.02.vcf.gz",
	"./data/rotavirus_rf.03.vcf.gz",
	"./data/rotavirus_rf.04.bcf"
	)
# print the number of variants for each vcf
for(f in vcfs) {
	cat(paste(f," ",count.variants(f),"\n"))
	}
```

**Output**:

```
./data/gnomad.exomes.r2.0.1.sites.bcf   50 
./data/rotavirus_rf.01.vcf   45 
./data/rotavirus_rf.02.vcf.gz   45 
./data/rotavirus_rf.03.vcf.gz   45 
./data/rotavirus_rf.04.bcf   45 
```


### Scanning the variants

**Code**:

```
# load rbcf
library(rbcf)

# create a function counting variants in a VCF
count.variants<-function(filename,predicate) {
	# we don't need the index for this file
	fp <- bcf.open(filename,FALSE)
	# error on opening
	if(is.null(fp)) return(-1)
	# number of variants
	n<-0
	# loop while we can read a variant
	while(!is.null(vc<-bcf.next(fp))) {
		# test the variant
		if(predicate(vc)) {
			# increment the count
			n<-n+1
			}
	}
	# dispose the vcf reader
	bcf.close(fp)
	# return the number of variant
	n
}

# A vcf
filename <- "./data/gnomad.exomes.r2.0.1.sites.bcf"
# filters
filters<-list(
	list("desc"="accept all","predicate"=function(ctx) {TRUE} ),
	list("desc"="accept none","predicate"=function(ctx) {FALSE} ),
	list("desc"="CHROM is '1'","predicate"=function(ctx) { variant.contig(ctx)=="1"} ),
	list("desc"="POS is even","predicate"=function(ctx) { (variant.pos(ctx)%%2)==1} ),
	list("desc"="PASS filter","predicate"=function(ctx) {!variant.is.filtered(ctx)} ),
	list("desc"="count(FILTER)>1","predicate"=function(ctx) {length(variant.filters(ctx))>1} ),
	list("desc"="FILTER contains SEGDUP","predicate"=function(ctx) {variant.has.filter(ctx,"SEGDUP")} ),
	list("desc"="SNP","predicate"=function(ctx) {variant.is.snp(ctx)} ),
	list("desc"="POS!=END","predicate"=function(ctx) { variant.pos(ctx)!=variant.end(ctx)} ),
	list("desc"="not diallelic","predicate"=function(ctx) {variant.nalleles(ctx)!=2} ),
	list("desc"="REF is 'A'","predicate"=function(ctx) {variant.reference(ctx)=="A"} ),
	list("desc"="any allele is 'A'","predicate"=function(ctx) {"A"  %in% variant.alleles(ctx)} ),
	list("desc"="any ALT allele is 'A'","predicate"=function(ctx) {"A"  %in% variant.alt.alleles(ctx)} ),
	list("desc"="No QUAL","predicate"=function(ctx) {!variant.has.qual(ctx)} ),
	list("desc"="variant has ID","predicate"=function(ctx) {variant.has.id(ctx)}),
	list("desc"="variant ID match 'rs1*' ","predicate"=function(ctx) {grepl("^rs1",variant.id(ctx))}),
	list("desc"="variant has INFO/AF_NFE","predicate"=function(ctx) {variant.has.attribute(ctx,"AF_NFE")}),
	list("desc"="variant has INFO/AF_NFE > 1E-5","predicate"=function(ctx) {variant.has.attribute(ctx,"AF_NFE") && length(which(variant.float.attribute(ctx,"AF_NFE") > 1E-5))>0}),
	list("desc"="Missense in PLEKHN1 (VEP)","predicate"=function(ctx) {
		# NO VEP annotation ?
		if(!variant.has.attribute(ctx,"CSQ")) return(FALSE);
		# get VEP annotation
		predictions <- variant.vep(ctx)
		# In SCN5A
		predictions <- predictions[which(predictions$SYMBOL=="PLEKHN1"),]
		# Consequence must contain missense
		predictions <- predictions[grep("missense_variant",predictions$Consequence),]
		nrow(predictions)>0
		})
	)

# count the variant for each filter
for(flt in filters) {
	print(paste(basename(filename)," filter:",flt[["desc"]]," count:",count.variants(filename,flt[["predicate"]]),"\n"))
	}

	
```

**Output**:

```
[1] "gnomad.exomes.r2.0.1.sites.bcf  filter: accept all  count: 50 \n"
[1] "gnomad.exomes.r2.0.1.sites.bcf  filter: accept none  count: 0 \n"
[1] "gnomad.exomes.r2.0.1.sites.bcf  filter: CHROM is '1'  count: 50 \n"
[1] "gnomad.exomes.r2.0.1.sites.bcf  filter: POS is even  count: 24 \n"
[1] "gnomad.exomes.r2.0.1.sites.bcf  filter: PASS filter  count: 48 \n"
[1] "gnomad.exomes.r2.0.1.sites.bcf  filter: count(FILTER)>1  count: 2 \n"
[1] "gnomad.exomes.r2.0.1.sites.bcf  filter: FILTER contains SEGDUP  count: 1 \n"
[1] "gnomad.exomes.r2.0.1.sites.bcf  filter: SNP  count: 47 \n"
[1] "gnomad.exomes.r2.0.1.sites.bcf  filter: POS!=END  count: 3 \n"
[1] "gnomad.exomes.r2.0.1.sites.bcf  filter: not diallelic  count: 8 \n"
[1] "gnomad.exomes.r2.0.1.sites.bcf  filter: REF is 'A'  count: 6 \n"
[1] "gnomad.exomes.r2.0.1.sites.bcf  filter: any allele is 'A'  count: 27 \n"
[1] "gnomad.exomes.r2.0.1.sites.bcf  filter: any ALT allele is 'A'  count: 21 \n"
[1] "gnomad.exomes.r2.0.1.sites.bcf  filter: No QUAL  count: 1 \n"
[1] "gnomad.exomes.r2.0.1.sites.bcf  filter: variant has ID  count: 34 \n"
[1] "gnomad.exomes.r2.0.1.sites.bcf  filter: variant ID match 'rs1*'   count: 2 \n"
[1] "gnomad.exomes.r2.0.1.sites.bcf  filter: variant has INFO/AF_NFE  count: 50 \n"
[1] "gnomad.exomes.r2.0.1.sites.bcf  filter: variant has INFO/AF_NFE > 1E-5  count: 14 \n"

NEWS.md
# 20200420

first release


R/rbcf.R
#' @return the version of htslib
htslib.version <-function()  {
	.Call("HtslibGetVersion");
	}

#' @return the version of rcbf
rcbf.version <-function()  {
        .Call("RBCFGetVersion");
        }

#' Open a VCF or a BCF file
#' 
#' fp<-bcf.open("my.vcf.gz",TRUE)
#' hts.close(fp)
#' fp<-bcf.open("my.bcf")
#' hts.close(fp)
#' 
#' @param filename the path to the vcf file
#' @param requireIndex load associated vcf index
#' @return the new VCF reader
bcf.open<-function(filename,requireIndex=TRUE)
	{
	stopifnot(is.character(filename))
	stopifnot(is.logical(requireIndex))
	.Call("RBcfFileOpen",filename,requireIndex);
	}

#' Close a VCF reader
#' 
#' fp<-bcf.open("my.vcf.gz",TRUE)
#' bcf.close(fp)
#' 
#' @param fp the vcf reader
#' @return true on success
bcf.close<-function(fp)
	{
	if(!is.null(fp)) .Call("RBcfFileClose",fp);
	}


#' Open a new VCF writer.
#' Must be closed with bcf.close
#' 
#' @param fp the vcf reader
#' @param fname the name of the output vcf file
#' @return the writer or NULL on failure
bcf.new.writer<-function(fp,fname="-")
	{
	.Call("RBcfNewWriter",fp,fname);
	}

#' Save a Variant in a VCF writer
#' 
#' @param fp the vcf reader
#' @param vc the variant
#' @return true on success
bcf.write.variant<-function(fp,vc)
	{
	.Call("RBcfFileWriteCtx",fp,vc);
	}

#' @param fp the vcf reader
#' @return the number of samples
bcf.nsamples<-function(fp)
	{
	.Call("RBcfNSamples",fp);
	}

#' @param fp the vcf reader
#' @return samples
bcf.samples<-function(fp)
	{
	.Call("RBcfSamples",fp);
	}

#'
#' @param fp the vcf reader
#' @param sn the sample name
bcf.sample2index<-function(fp,sn) {
	sapply(sn,function(S) {
		.Call("BcfConvertSampleToIndex0",fp,S) + 1 
		})
	}
#' list the indexed chromosomes 
#' @param fp the vcf reader
#' @return a list of chromosome
bcf.chromosomes<-function(fp)
	{
	.Call("RBcfSeqNames",fp)
	}

#' alias of bcf.chromosomes
#' @param fp the vcf reader
#' @return a list of chromosome
bcf.contigs<-function(fp)
	{
	bcf.chromosomes(fp)
	}

#' @param fp the vcf reader
#' @param idx the 1-based index
#' @return the idx-th sample (1-based)
bcf.sample.at<-function(fp,idx)
	{
	stopifnot(idx>0)
	.Call("RBcfSampleAtIndex0",fp,idx-1)
	}

#' Close a VCF reader
#' 
#' fp<-bcf.open("my.vcf.gz",TRUE)
#' bcf.close(fp)
#'
#' @param fp the vcf reader
#' @return true on success
bcf.close<-function(fp)
	{
	.Call("RBcfFileClose",fp);
	}

#' get dictionary from a VCF reader.
#'
#' 
#' fp<-bcf.open("my.vcf.gz",TRUE)
#' dict<-bcf.dictionary(fp)
#' print(dict)
#'      chrom size
#' RF01  RF01 3302
#' RF02  RF02 2687
#' RF03  RF03 2592
#' RF04  RF04 2362
#' RF05  RF05 1579
#' RF06  RF06 1356
#' RF07  RF07 1074
#' RF08  RF08 1059
#' RF09  RF09 1062
#' RF10  RF10  751
#' RF11  RF11  666
#' 
#' @param fp the vcf reader
#' @return the vcf dictionary as a table
bcf.dictionary<-function(fp)
	{
	.Call("RBcfHeaderDict",fp);
	}


#' prepare the VCF reader for a new vcf iteration over a given interval.
#' VCF reader must be associated and opened with a valid index.
#' 
#' @param fp the vcf reader
#' @param interval the genomic interval to be scanned
#' @param collect if TRUE return a list of variants in the region or NULL on failure
#' @return TRUE on success or a list of variant if collect=TRUE
#' 
bcf.query<-function(fp,interval,collect=FALSE) {
	stopifnot(is.character(interval))
	ret<-.Call("RBcfQueryRegion",fp,interval);
	
	if(collect) {
		if(!ret) return(NULL);
		# initialize
		variants <-list()
		n<- 1
		# loop while we can read a variant
			while(!is.null(vc<-bcf.next(fp))) {
				#append
				variants[[n]] = vc
				n <- n + 1
				}
		return(variants)
		}
	ret
	}
	
#' read the next Variant Context in the VCF reader
#' 
#' 
#' fp <- bcf.open("in.vcf.gz")
#' bcf.query(fp,"RF02:1-1000")
#' while(!is.null(vc<-bcf.next(fp))) {
#'      ## do something with vc
#'      }
#' bcf.close(fp)
#'
#' @param fp the vcf reader
#' @return an opaque vcf context or NULL 
#
bcf.next<-function(fp) {
	.Call("RBcfNextLine",fp);
	}

#' get the numeric index (tid) of the chromosome for this variant
#' 
#' @param vc the variant
#' @return the numeric index for this variant
#
variant.tid<-function(vc) {
	.Call("RBcfCtxRid",vc);
	}
#' get chromosome name for this variant
#' 
#' @param vc the variant
#' @return the chromosome name
#
variant.contig<-function(vc) {
	.Call("RBcfCtxSeqName",vc);
	}
#' alias of variant.contig
#' 
#' @param vc the variant
#' @return the chromosome name
variant.chrom<-function(vc) {
	variant.contig(vc)
	}
#' get chromosome POS for this variant
#' 
#' @param vc the variant
#' @return the starting position for this variant
#
variant.pos<-function(vc) {
	.Call("RBcfCtxPos",vc);
	}
#' alias of variant.pos
#' 
#' @param vc the variant
#' @return the starting position for this variant
variant.start<-function(vc) {
	variant.pos(vc)
	}

#' return whether variant have got an ID
#' 
#' @param vc the variant
#' @return the TRUE if variant have got an ID
variant.has.id<-function(vc) {
	.Call("RBcfCtxHasId",vc);
	}
	
#' return the variant ID
#' 
#' @param vc the variant
#' @return the variant ID or NULL
variant.id<-function(vc) {
	.Call("RBcfCtxId",vc);
	}

#' get chromosome END for this variant
#' 
#' @param vc the variant
#' @return the END position for this variant
#
variant.end<-function(vc) {
	.Call("RBcfCtxEnd",vc);
	}

#' alias if variant.end
#' 
#' @param vc the variant
#' @return the END position for this variant
#
variant.stop<-function(vc) {
	variant.end(vc);
	}


#' get the number of alleles for this variant
#' 
#' @param vc the variant
#' @return the number of alleles for this variant
#
variant.nalleles<-function(vc) {
	.Call("RBcfCtxNAlleles",vc);
	}

#' get the alleles for this variant
#' 
#' @param vc the variant
#' @return the alleles for this variant
#
variant.alleles<-function(vc) {
	.Call("RBcfCtxAlleles",vc);
	}

#' get the reference allele for this variant
#' 
#' @param vc the variant
#' @return the reference allele for this variant
#
variant.reference<-function(vc) {
	.Call("RBcfCtxReference",vc);
	}

#' get the alternate alleles for this variant
#' 
#' @param vc the variant
#' @return the alternate alleles for this variant
#
variant.alt.alleles<-function(vc) {
	.Call("RBcfCtxAlternateAlleles",vc);
	}


#' test if QUAL is available for this variant
#' 
#' @param vc the variant
#' @return QUAL is available for this variant
#
variant.has.qual<-function(vc) {
	.Call("RBcfCtxHasQual",vc);
	}

#' test the QUAL for this variant
#' 
#' @param vc the variant
#' @return QUAL or NULL
#
variant.qual<-function(vc) {
	.Call("RBcfCtxQual",vc);
	}
	

#' test is variant is filtered
#' 
#' @param vc the variant
#' @return  variant is filtered
#
variant.is.filtered<-function(vc) {
	.Call("RBcfCtxFiltered",vc);
	}

#' return the FILTERS for this variant
#' 
#' @param vc the variant
#' @return the filters
#
variant.filters<-function(vc) {
	.Call("RBcfCtxFilters",vc);
	}


#' @param vc the variant
#' @param fn filter name
#' @return true if variant is filtered with 'fn'
#
variant.has.filter<-function(vc,fn) {
	.Call("VariantHasFilter",vc,fn);
	}


#' return the types for this variant (as defined in htslib)
#' 
#' @param vc the variant
#' @return the types for this variant
#
variant.types<-function(vc) {
	.Call("RBcfCtxVariantTypes",vc);
	}


#' return true if variant is a SNP
#' 
#' @param vc the variant
#' @return true if the variant is a SNP
#
variant.is.snp<-function(vc) {
	.Call("RBcfCtxVariantIsSnp",vc);
	}

#' max ploidy for this variant
#' 
#' @param vc the variant
#' @return max ploidy
#
variant.max.ploidy<-function(vc) {
	.Call("RBcfCtxVariantMaxPloidy",vc);
	}

#' @param vc the variant
#' @return the number of samples/genotypes for this variant
variant.nsamples <-function(vc) {
	.Call("VariantNSamples",vc);
	}

#' @param vc the variant
#' @return the number of samples/genotypes for this variant
variant.genotypes <-function(vc) {
	sapply(1:variant.nsamples(vc),function(idx) {
		variant.genotype(vc,idx)
		})
        }


#' return the genotype for a variant
#' 
#' @param vc the variant
#' @param nameOrIdx the sample name (slower) or the 1-based sample index 
#' @return the given genotype
#
variant.genotype<-function(vc,nameOrIdx) {
	if(is.numeric(nameOrIdx)) {
		stopifnot(nameOrIdx>0)
		nameOrIdx<-as.integer(nameOrIdx)-1
		}
	else {
		stopifnot(is.character(nameOrIdx))
		}
	.Call("VariantGetGenotype",vc,nameOrIdx);
	}

#' return the 0-based alleles indexes for the given genotype
#' 
#' @param gt the genotype
#' @return max ploidy
#
genotype.alleles.idx0 <- function(gt) {
	.Call("RBcfCtxVariantGtAllelesIndexes0",gt);
}

#' return the 0-based allele indizes for all genotypes of the variant
#' 
#' The returned vector contains the allele-indizes for all genotypes in sample order
#' 
#' For example, for a 3 sample VCF on a diploid variant, the returned vector could be 
#' ```
#'   GT1  GT2  GT3
#' c(0,1, 0,0, 1,1)
#' ```
#' 
#' @param vc the variant
#' @return an integer vector of length `ploidy * n_samples`
variant.genotypes.allele.idx0 <- function(vc) {
	.Call("RBcfCtxVariantAllGtAllelesIndexes0", vc);
	}

#' return the genotype strings for all genotypes of the variant
#' 
#' The returned vector contains the genotype strings in sample order.
#' 
#' For example, for a 3 sample VCF on a diploid variant, the returned vector could be 
#' ```
#'    GT1    GT2    GT3
#' c("0/1", "0/0", "1/1")
#' ```
#' 
#' @param vc the variant
#' @return a character vector of length `n_samples`
variant.genotypes.allele.strings <-function(vc) {
	.Call("RBcfCtxVariantAllGtStrings", vc);
	}

#' @param gt the genotype
#' @return the number of alleles for the genotypes
#
genotype.ploidy <-function(gt) {
	alleles<-genotype.alleles.idx0(gt)
	v = 0;
	if(!is.null(alleles)) v= length(alleles)
	v
	}

#' @param gt the genotype
#' @return TRUE if genotype is diploid and all alleles are reference
#
genotype.homref <-function(gt) {
	alleles<-genotype.alleles.idx0(gt)
	!is.null(alleles) && length(alleles)==2 && alleles[1]==0 && alleles[2]==0 
	}

#' @param gt the genotype
#' @return TRUE if genotype is diploid and heterozygous
#
genotype.het <-function(gt) {
	alleles<-genotype.alleles.idx0(gt)
	!is.null(alleles) && length(alleles)==2 && alleles[1]!=alleles[2] && alleles[1]>=0 && alleles[2]>=0
	}

#' @param gt the genotype
#' @return TRUE if genotype is diploid and homozygous on alt allele
#
genotype.homvar <-function(gt) {
	alleles<-genotype.alleles.idx0(gt)
	!is.null(alleles) && length(alleles)==2 && alleles[1]>0 && alleles[2]>0 && alleles[1]==alleles[2]
	}
	
#' @param gt the genotype
#' @return TRUE if genotype is diploid and heterozygous and doesn't contains the alt allele
#
genotype.hetnonref  <-function(gt) {
	alleles<-genotype.alleles.idx0(gt)
	!is.null(alleles) && length(alleles)==2 && alleles[1]!=alleles[2] && alleles[1]>0 && alleles[2]>0
	}

#' @param gt the genotype
#' @return TRUE if genotypes contains no allele '' or any is no call '.'
genotype.nocall  <-function(gt) {
	alleles<-genotype.alleles.idx0(gt)
	is.null(alleles) || length(alleles)==0 || -1 %in% alleles
	}

#' @param gt the genotype
#' @return TRUE if genotype is phased
#
genotype.phased  <-function(gt) {
	.Call("RBcfCtxVariantGtPhased",gt);
	}

#' @param gt the genotype
#' @return the name of the sample associated to the genotype
genotype.sample  <-function(gt) {
	.Call("GenotypeSample",gt)
	}


#' @param vc the variant
#' @param att the INFO/Attribute
#' @return true if variant has INFO attribute
variant.has.attribute <-function(vc,att) {
	stopifnot(is.character(att))
	.Call("VariantHasAttribute",vc,att)
	}


#' @param vc the variant
#' @param att the INFO/Attribute
#' @param split split strings using commas
#' @return the the INFO attribute for the given key.
variant.string.attribute <-function(vc,att, split = TRUE) {
	stopifnot(is.character(att))
	s <-.Call("VariantStringAttribute",vc,att)
	if( !is.null(s) && length(s)>0 && split) {
		s <- unlist(strsplit(s, split=","))
		}
	s
	}

#' @param vc the variant
#' @param att the INFO/Attribute
#' @return the the INFO attribute for the given key.
variant.int.attribute <-function(vc,att) {
	.Call("VariantIntAttribute",vc,att)
	}

#' @param vc the variant
#' @param att the INFO/Attribute
#' @return the the INFO attribute for the given key.
variant.float.attribute <-function(vc,att) {
	.Call("VariantFloatAttribute",vc,att)
	}

#' @param vc the variant
#' @param att the INFO/Attribute
#' @return the the INFO attribute for the given key.
variant.flag.attribute <-function(vc,att) {
	.Call("VariantFlagAttribute",vc,att)
	}


#' @param fp the vcf reader
#' @return a table of filters
bcf.filters <-function(fp) {
	 .Call("BcfFilterTable",fp)
}

#' @param fp the vcf reader
#' @return a table of filters
bcf.infos <-function(fp) {
	 .Call("BcfInfoTable",fp)
}

#' @param fp the vcf reader
#' @return a table of filters
bcf.formats <-function(fp) {
	 .Call("BcfFormatTable",fp)
	}

#' @param vc the variant
#' @return the list INFOs for this variant
variant.info.ids <-function(vc) {
	.Call("VariantGetInfoKeySet",vc)
	}

#' @param vc the variant
#' @return the list FORMATs for this variant
variant.format.ids <-function(vc) {
	.Call("VariantGetFormatKeySet",vc)
	}
#' @param vc the variant
#' @return VEP table for this variant or NULL
variant.vep <-function(vc) {
	.Call("VariantVepTable",vc)
	}
#' @param vc the variant
#' @return SNPEFF table for this variant or NULL
variant.snpeff <-function(vc) {
	.Call("VariantSnpEffTable",vc)
	}



#' Return a specific FORMAT flag value on all genotypes
#' 
#' The returned vector contains the flags by sample and attribute number (if the attribute comprises multiple flags).
#' 
#' For example, for a 3 sample VCF with a flag having a single logical, the returned vector could be 
#' ```
#'    GT1    GT2   GT3
#' c(TRUE, FALSE, TRUE)
#' ```
#' @param vc the variant
#' 
#' @return vector of logicals containing attribute values for all genotypes
#' 
#' @seealso 
#'    \link{variant.genotypes.int.attribute}, 
#'    \link{variant.genotypes.float.attribute}
#'    
variant.genotypes.flag.attribute <- function(vc, att) {
	as.logical(.Call("VariantGenotypesFlagAttribute", vc, att))
}

#' Return a specific FORMAT integer value on all genotypes
#' 
#' The returned vector contains the values by sample and attribute number (if the attribute comprises multiple integer values).
#' 
#' For example, for a 3 sample VCF extracting the allelic read depth (AD) on a singleton variant, the results could look like: 
#' ```
#'   GT1-REF, GT1-ALT, GT2-REF, GT2-ALT, GT3-REF, GT3-ALT
#' c(     10,     100,      25,      94,      45,       7)
#' ```
#' 
#' @param vc the variant
#' 
#' @return vector of numerics containing attribute values for all genotypes
#' 
#' @seealso 
#'    \link{variant.genotypes.flag.attribute}, 
#'    \link{variant.genotypes.float.attribute}
#'    
variant.genotypes.int.attribute <- function(vc, att) {
	.Call("VariantGenotypesInt32Attribute", vc, att)
}

#' Return a specific FORMAT numeric value on all genotypes
#' 
#' The returned vector contains the values by sample and attribute number (if the attribute comprises multiple integer values).
#' 
#' For example, for a 3 sample VCF extracting the allelic read fraction (AF) on a multi-allelic variant, the results could look like: 
#' ```
#'   GT1-ALT1, GT1-ALT2, GT2-ALT1, GT2-ALT2, GT3-ALT1, GT3-ALT2
#' c(     0.9,      0.1,      0.6,      0.1,      0.1,      0.4)
#' ```
#' 
#' @param vc the variant
#' 
#' @return vector of numerics containing attribute values for all genotypes
#' 
#' @seealso 
#'    \link{variant.genotypes.flag.attribute}, 
#'    \link{variant.genotypes.float.attribute}
#'
variant.genotypes.float.attribute <- function(vc, att) {
	.Call("variant.genotypes.int.attribute", vc, att)
}

#' @param gt the genotype
#' @param att the key
#' @return the values for this key+genotype
#
genotype.int.attribute <-function(gt,att) {
	.Call("GenotypeInt32Attribute",gt,att)
	}


#' @param gt the genotype
#' @param att the key
#' @return a String for this flag+genotype or NULL
#
genotype.string.attribute <-function(gt,att) {
	.Call("GenotypeStringAttribute",gt,att)
	}

#' @param gt the genotype
#' @return TRUE if genotype is filtered (FORMAT/FT is set)
#
genotype.filtered <-function(gt) {
	flt<- genotype.string.attribute(gt,"FT")
	!(is.null(flt) || length(flt)==0 || flt==".")
	}

#' @param gt the genotype
#' @return the DP or -1
genotype.dp <-function(gt) {
	v<-genotype.int.attribute(gt,"DP")
	if(length(v)!=1) return(-1)
	v[1]	
	}
#' @param gt the genotype
#' @return TRUE if DP is available 
genotype.has.dp <-function(gt) {
	genotype.dp(gt)>=0	
	}
#' @param gt the genotype
#' @return the GQ
genotype.gq <-function(gt) {
	v<-genotype.int.attribute(gt,"GQ")
	if(length(v)!=1) return(-1)
	v[1]	
	}
#' @param gt the genotype
#' @return TRUE if GQ is available 
genotype.has.gq <-function(gt) {
	genotype.gq(gt)>=0	
	}
#' @param gt the genotype
#' @return PL or empty vector
genotype.pl <-function(gt) {
	genotype.int.attribute(gt,"PL")
	}
#' @param gt the genotype
#' @return TRUE if PL is available 
genotype.has.pl<-function(gt) {
	length(genotype.pl(gt))>0	
	}

#' @param gt the genotype
#' @return AD or rempty vector
genotype.ad <-function(gt) {
	genotype.int.attribute(gt,"AD")
	}

#' @param gt the genotype
#' @return TRUE if AD is available 
genotype.has.ad <-function(gt) {
	length(genotype.ad(gt))>0	
	}


#' @param gt the genotype
#' @param att the key
#' @return the values for this key+genotype
#
genotype.float.attribute <-function(gt,att) {
	.Call("GenotypeFloatAttribute",gt,att)
	}
